### PR TITLE
Amls 5877 - WYN content changes

### PR DESCRIPTION
**Pull request by matthewharris1**
There is a tonne of if logic in this bit of work so will require extensive testing however what I have checked manually looks good and there is unit tests for each line of content. 

## Checklist

- [ ] Breaking change (fix or feature that would cause existing functionality to change).
- [x] Build passing.
- [x] Unit tests have full coverage.
- [ ] Unit test approach and implementation has been reviewed with QA (testers).
- [x] Requires acceptance test run.
- [x] Appropriate labels added.
- [x] RELEASE NOTES ADDED.

### CODE REVIEW

File: app/controllers/businessactivities/WhatYouNeedController.scala [hunk 1]
```diff
@@ -34,6 +42,13 @@ import scala.concurrent.Future
...
   def get = authAction.async {
     implicit request =>
-      Future.successful(Ok(what_you_need(routes.InvolvedInOtherController.get().url)))
+      dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key) map { businessMatching =>
+        (for {
+          bm <- businessMatching
+          ba <- bm.activities
+        } yield {
+          Ok(what_you_need(routes.InvolvedInOtherController.get().url, Some(ba)))
+        }) getOrElse Redirect(controllers.routes.RegistrationProgressController.get())
```
Comment (ajchilts) [inline app/controllers/businessactivities/WhatYouNeedController.scala:51]: Just interested. Why Registration Progress Controller if we fail to get the activities? 

Do we want to do this? Is this error recoverable? If not we likely want to raise an exception such that we can investigate. If we encounter a non recoverable error such as failing to locate business matching in the cache then sending the user back to the progress page might be confusing? The user will keep trying to get into the section and could get stuck in a loop without an exception being thrown therefore giving us nothing to investigate. 

If we need this data to be able to proceed then should we raise an error if this data can't be located? 

The alternative is to load a default if the data can't be retrieved.

File: app/controllers/renewal/WhatYouNeedController.scala [hunk 2]
```diff
@@ -15,29 +15,41 @@
...
+                                       renewalService: RenewalService,
+                                       val cc: MessagesControllerComponents)(implicit ec: ExecutionContext) extends AmlsBaseController(ds, cc) {
   def get = authAction.async {
     implicit request =>
-      renewalService.getSection(request.credId) map {
-        case Section(_,NotStarted | Started,_,_) => Ok(what_you_need())
-        case _ => Redirect(controllers.routes.RegistrationProgressController.get())
-      }
+        (for {
+          cache <- OptionT(dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key))
```
Comment (ajchilts) [inline app/controllers/renewal/WhatYouNeedController.scala:42]: Can we please rename this to `BusinessMatching` or `bm` as this is only one section of the cache?

File: app/controllers/renewal/WhatYouNeedController.scala [hunk 3]
```diff
@@ -15,29 +15,41 @@
...
+          ba <- OptionT.fromOption[Future](cache.activities)
+          section <- OptionT.liftF(getSection(renewalService, request.credId, Some(ba), cache.msbServices))
+        } yield {
+          section
+        }).getOrElse(Redirect(controllers.routes.RegistrationProgressController.get()))
+  }
+  def getSection(renewalService: RenewalService, credId: String, ba: Option[BusinessActivities], msbActivities: Option[BusinessMatchingMsbServices])(implicit request: Request[_]) = {
```
Comment (ajchilts) [inline app/controllers/renewal/WhatYouNeedController.scala:49]: Maybe split this over multiple lines?

File: app/views/renewal/what_you_need.scala.html [hunk 4]
```diff
@@ -17,7 +17,16 @@
...
-@()(implicit request: Request[_],m:Messages, lang: Lang, appConfig: ApplicationConfig)
+@import models.businessmatching.BusinessActivities
+@import models.businessmatching.ArtMarketParticipant
+@import models.businessmatching.MoneyServiceBusiness
+@import models.businessmatching.BusinessMatchingMsbServices
+@import models.businessmatching.TransmittingMoney
+@import models.businessmatching.CurrencyExchange
+@import models.businessmatching.AccountancyServices
+@import models.businessmatching.HighValueDealing
+@import models.businessmatching.ForeignExchange
```
Comment (ajchilts) [inline app/views/renewal/what_you_need.scala.html:28]: Would `@import models.businessmatching._` work better here?

File: app/views/renewal/what_you_need.scala.html [hunk 5]
```diff
@@ -37,7 +46,32 @@
...
+                    <li>@Messages("renewal.whatyouneed.line_8")</li>
+                }
+                @if(ba.hasBusinessOrAdditionalActivity(MoneyServiceBusiness) && msb.msbServices.contains(CurrencyExchange)) {
+                    <li>@Messages("renewal.whatyouneed.line_9")</li>
+                    <li>@Messages("renewal.whatyouneed.line_10")</li>
+                    <li>@Messages("renewal.whatyouneed.line_11")</li>
+                }
+                @if(ba.hasBusinessOrAdditionalActivity(MoneyServiceBusiness) && msb.msbServices.contains(ForeignExchange)) {
+                    <li>@Messages("renewal.whatyouneed.line_12")</li>
+                }
```
Comment (ajchilts) [inline app/views/renewal/what_you_need.scala.html:66]: Could be simplified? 

```
@if(ba.hasBusinessOrAdditionalActivity(MoneyServiceBusiness) {

               @if(msb.msbServices.contains(TransmittingMoney)) {
                    <li>@Messages("renewal.whatyouneed.line_6")</li>
                    <li>@Messages("renewal.whatyouneed.line_7")</li>
                    <li>@Messages("renewal.whatyouneed.line_8")</li>
                }

                @if(msb.msbServices.contains(CurrencyExchange)) {
                    <li>@Messages("renewal.whatyouneed.line_9")</li>
                    <li>@Messages("renewal.whatyouneed.line_10")</li>
                    <li>@Messages("renewal.whatyouneed.line_11")</li>
                }

                @if(msb.msbServices.contains(ForeignExchange)) {
                    <li>@Messages("renewal.whatyouneed.line_12")</li>
                }
}
```

File: app/views/tradingpremises/what_you_need.scala.html [hunk 6]
```diff
@@ -17,46 +17,68 @@
...
+        @if(ba.businessActivities.size > 1) {
+            <li>@Messages("tradingpremises.whatyouneed.requiredinfo.text.4")</li>
+        }
+        @msbServices.map { msb =>
+            @if(ba.hasBusinessOrAdditionalActivity(MoneyServiceBusiness) && msb.msbServices.size > 1) {
+                <li>@Messages("tradingpremises.whatyouneed.requiredinfo.text.5")</li>
+            }
+        }
+
+
```
Comment (ajchilts) [inline app/views/tradingpremises/what_you_need.scala.html:57]: Extra whitespace

File: conf/messages [hunk 7]
```diff
@@ -1404,12 +1406,16 @@ businessdetails.lettersaddress.title = Is this the right address to send you let
...
-tradingpremises.whatyouneed.agents.desc.1 = If you use agents, we also need to know where your agents trade from.
-tradingpremises.whatyouneed.agents.desc.2 = An agent is an individual or business that works on behalf of your business to provide your services. An employee is not an agent.
+tradingpremises.whatyouneed.requiredinfo.text.2 = the date you started or will start trading from these premises
+tradingpremises.whatyouneed.requiredinfo.text.3 = if it is a residential address
+tradingpremises.whatyouneed.requiredinfo.text.4 = the services your business provides at these premises
+tradingpremises.whatyouneed.requiredinfo.text.5 = which money service business activities your business does at these premises
+tradingpremises.whatyouneed.requiredinfo.text.6 = company name and registration number
+tradingpremises.whatyouneed.requiredinfo.text.7 = name and date of birth
+tradingpremises.whatyouneed.requiredinfo.text.8 = business partners
+tradingpremises.whatyouneed.agents.sub.heading = Agents’ trading premises
```
Comment (ajchilts) [inline conf/messages:1416]: Is that punctuation mark in the right place?

File: test/controllers/businessactivities/WhatYouNeedControllerSpec.scala [hunk 8]
```diff
@@ -85,6 +103,18 @@ class WhatYouNeeControllerSpec extends AmlsSpec with MockitoSugar with ScalaFutu
           doc.getElementById("ba-whatyouneed-button").attr("href") mustBe routes.InvolvedInOtherController.get().url
         }
       }
+      "Redirect to registration progress page" when {
```
Comment (ajchilts) [inline test/controllers/businessactivities/WhatYouNeedControllerSpec.scala:106]: As stated in the controller. Do we want to do this? Is this error recoverable? If not we likely want to raise an exception such that we can investigate. If we encounter a non recoverable error such as failing to locate business matching in the cache then sending the user back to the progress page might be confusing? The user will keep trying to get into the section and could get stuck in a loop without an exception being thrown therefore giving us nothing to investigate. 

If we need this data to be able to proceed then should we raise an error if this data can't be located? 

The alternative is to load a default if the data can't be retrieved.

File: app/controllers/renewal/WhatYouNeedController.scala [hunk 9]
```diff
@@ -15,29 +15,41 @@
...
-        case Section(_,NotStarted | Started,_,_) => Ok(what_you_need())
-        case _ => Redirect(controllers.routes.RegistrationProgressController.get())
-      }
+        (for {
+          cache <- OptionT(dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key))
+          ba <- OptionT.fromOption[Future](cache.activities)
+          section <- OptionT.liftF(getSection(renewalService, request.credId, Some(ba), cache.msbServices))
+        } yield {
+          section
+        }).getOrElse(Redirect(controllers.routes.RegistrationProgressController.get()))
```
Comment (ajchilts) [inline app/controllers/renewal/WhatYouNeedController.scala:47]: As stated above. Do we want to do this? Is this error recoverable? If not we likely want to raise an exception such that we can investigate. If we encounter a non recoverable error such as failing to locate business matching in the cache then sending the user back to the progress page might be confusing? The user will keep trying to get into the section and could get stuck in a loop without an exception being thrown therefore giving us nothing to investigate. 

If we need this data to be able to proceed then should we raise an error if this data can't be located? 

The alternative is to load a default if the data can't be retrieved.

File: app/controllers/responsiblepeople/WhatYouNeedController.scala [hunk 10]
```diff
@@ -17,22 +17,33 @@
...
     authAction.async {
       implicit request =>
-        Future.successful(Ok(what_you_need(index, flow)))
+        dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key) map { businessMatching =>
+          (for {
+            bm <- businessMatching
+            ba <- bm.activities
+          } yield {
+            Ok(what_you_need(index, flow, Some(ba)))
+          }) getOrElse Redirect(controllers.routes.RegistrationProgressController.get())
```
Comment (ajchilts) [inline app/controllers/responsiblepeople/WhatYouNeedController.scala:46]: As above.

File: app/controllers/tradingpremises/WhatYouNeedController.scala [hunk 11]
```diff
@@ -34,10 +34,12 @@ class WhatYouNeedController @Inject()(val dataCacheConnector: DataCacheConnector
...
-      dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key) map {
-        response =>
-        Ok(what_you_need(index, ControllerHelper.isMSBSelected(response)))
-      }
+      dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key) map { businessMatching =>
+        (for {
+          bm <- businessMatching
+          ba <- bm.activities
+        } yield { Ok(what_you_need(index, Some(ba), bm.msbServices))
+      })getOrElse  Redirect(controllers.routes.RegistrationProgressController.get())
```
Comment (ajchilts) [inline app/controllers/tradingpremises/WhatYouNeedController.scala:42]: As above

File: app/views/tradingpremises/what_you_need.scala.html [hunk 12]
```diff
@@ -17,46 +17,68 @@
...
+        @if(ba.hasBusinessOrAdditionalActivity(MoneyServiceBusiness)) {
+            <h3 class="heading-medium">@Messages("tradingpremises.whatyouneed.agents.sub.heading")</h3>
+
+            <p>@Messages("tradingpremises.whatyouneed.agents.desc.1")</p>
+
+            <p>@Messages("tradingpremises.whatyouneed.agents.desc.2")</p>
+
+
```
Comment (ajchilts) [inline app/views/tradingpremises/what_you_need.scala.html:65]: Extra whitespace

File: app/views/renewal/what_you_need.scala.html (hunk 5 above)
Comment (matthewharris1) [reply app/views/renewal/what_you_need.scala.html:66]: How?

File: conf/messages (hunk 7 above)
Comment (matthewharris1) [reply conf/messages:1416]: yes

File: app/controllers/businessactivities/WhatYouNeedController.scala (hunk 1 above)
Comment (matthewharris1) [reply app/controllers/businessactivities/WhatYouNeedController.scala]: Done

File: app/controllers/renewal/WhatYouNeedController.scala (hunk 2 above)
Comment (matthewharris1) [reply app/controllers/renewal/WhatYouNeedController.scala]: Done

File: app/controllers/renewal/WhatYouNeedController.scala (hunk 3 above)
Comment (matthewharris1) [reply app/controllers/renewal/WhatYouNeedController.scala]: Done

File: app/views/renewal/what_you_need.scala.html (hunk 4 above)
Comment (matthewharris1) [reply app/views/renewal/what_you_need.scala.html]: Done

File: app/views/tradingpremises/what_you_need.scala.html (hunk 6 above)
Comment (matthewharris1) [reply app/views/tradingpremises/what_you_need.scala.html]: Done

File: test/controllers/businessactivities/WhatYouNeedControllerSpec.scala (hunk 8 above)
Comment (matthewharris1) [reply test/controllers/businessactivities/WhatYouNeedControllerSpec.scala]: Done

File: app/controllers/renewal/WhatYouNeedController.scala (hunk 9 above)
Comment (matthewharris1) [reply app/controllers/renewal/WhatYouNeedController.scala]: Done

File: app/controllers/responsiblepeople/WhatYouNeedController.scala (hunk 10 above)
Comment (matthewharris1) [reply app/controllers/responsiblepeople/WhatYouNeedController.scala]: Done

File: app/controllers/tradingpremises/WhatYouNeedController.scala (hunk 11 above)
Comment (matthewharris1) [reply app/controllers/tradingpremises/WhatYouNeedController.scala]: Done

File: app/views/tradingpremises/what_you_need.scala.html (hunk 12 above)
Comment (matthewharris1) [reply app/views/tradingpremises/what_you_need.scala.html]: Done

File: app/controllers/renewal/WhatYouNeedController.scala [hunk 13]
```diff
@@ -15,29 +15,41 @@
...
-                                      renewalService: RenewalService,
-                                      val cc: MessagesControllerComponents)(implicit ec: ExecutionContext) extends AmlsBaseController(ds, cc) {
-
+class WhatYouNeedController @Inject()(
+                                       val dataCacheConnector: DataCacheConnector,
+                                       val authAction: AuthAction,
+                                       val ds: CommonPlayDependencies,
+                                       renewalService: RenewalService,
+                                       val cc: MessagesControllerComponents)(implicit ec: ExecutionContext) extends AmlsBaseController(ds, cc) {
   def get = authAction.async {
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/controllers/renewal/WhatYouNeedController.scala:40]: Done

File: app/controllers/renewal/WhatYouNeedController.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/controllers/renewal/WhatYouNeedController.scala]: Done

File: app/controllers/responsiblepeople/WhatYouNeedController.scala [hunk 14]
```diff
@@ -17,22 +17,33 @@
...
+
 import scala.concurrent.ExecutionContext.Implicits.global
 import views.html.responsiblepeople._
 
 import scala.concurrent.Future
 
 class WhatYouNeedController @Inject () (
+                                       val dataCacheConnector: DataCacheConnector,
                                        authAction: AuthAction,
                                        val ds: CommonPlayDependencies,
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/controllers/responsiblepeople/WhatYouNeedController.scala]: Done

File: app/views/businessactivities/what_you_need.scala.html [hunk 15]
```diff
@@ -32,20 +34,21 @@
...
+            <li>@Messages("businessactivities.whatyouneed.line_2")</li>
+            <li>@Messages("businessactivities.whatyouneed.line_3")</li>
+            <li>@Messages("businessactivities.whatyouneed.line_4")</li>
+            <li>@Messages("businessactivities.whatyouneed.line_5")</li>
+            <li>@Messages("businessactivities.whatyouneed.line_6")</li>
+            <li>@Messages("businessactivities.whatyouneed.line_7")</li>
+            <li>@Messages("businessactivities.whatyouneed.line_8")</li>
+            <li>@Messages("businessactivities.whatyouneed.line_9")</li>
+            <li>@Messages("businessactivities.whatyouneed.line_10")</li>
+            @if(!ba.hasBusinessOrAdditionalActivity(AccountancyServices)){<li>@Messages("businessactivities.whatyouneed.line_11")</li>}
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/views/businessactivities/what_you_need.scala.html]: Done

File: app/views/responsiblepeople/what_you_need.scala.html [hunk 16]
```diff
@@ -17,7 +17,10 @@
 @import include._
 @import forms2._
 @import config.{ ApplicationConfig}
-@(index: Int, flow: Option[String] = None)(implicit request: Request[_],m:Messages, lang: Lang, appConfig: ApplicationConfig)
+@import models.businessmatching.BusinessActivities
+@import models.businessmatching.MoneyServiceBusiness
+@import models.businessmatching.TrustAndCompanyServices
+@(index: Int, flow: Option[String] = None, activities: Option[BusinessActivities])(implicit request: Request[_],m:Messages, lang: Lang, appConfig: ApplicationConfig)
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/views/responsiblepeople/what_you_need.scala.html:24]: Done

File: app/views/businessactivities/what_you_need.scala.html [hunk 17]
```diff
@@ -17,7 +17,9 @@
 @import include._
 @import forms2._
 @import config.{ ApplicationConfig}
-@(nextPageUrl: String)(implicit request: Request[_],m:Messages, lang: Lang, appConfig: ApplicationConfig)
+@import models.businessmatching.BusinessActivities
+@import models.businessmatching.AccountancyServices
+@(nextPageUrl: String, activities: Option[BusinessActivities])(implicit request: Request[_],m:Messages, lang: Lang, appConfig: ApplicationConfig)
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/views/businessactivities/what_you_need.scala.html:23]: Done

File: app/views/renewal/what_you_need.scala.html [hunk 18]
```diff
@@ -17,7 +17,16 @@
...
+@import models.businessmatching.BusinessActivities
+@import models.businessmatching.ArtMarketParticipant
+@import models.businessmatching.MoneyServiceBusiness
+@import models.businessmatching.BusinessMatchingMsbServices
+@import models.businessmatching.TransmittingMoney
+@import models.businessmatching.CurrencyExchange
+@import models.businessmatching.AccountancyServices
+@import models.businessmatching.HighValueDealing
+@import models.businessmatching.ForeignExchange
+@(activities: Option[BusinessActivities], msbActivities: Option[BusinessMatchingMsbServices])(implicit request: Request[_],m:Messages, lang: Lang, appConfig: ApplicationConfig)
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/views/renewal/what_you_need.scala.html:22]: Done

File: app/controllers/tradingpremises/WhatYouNeedController.scala [hunk 19]
```diff
@@ -34,10 +34,12 @@ class WhatYouNeedController @Inject()(val dataCacheConnector: DataCacheConnector
...
-      dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key) map {
-        response =>
-        Ok(what_you_need(index, ControllerHelper.isMSBSelected(response)))
-      }
+      dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key) map { businessMatching =>
+        (for {
+          bm <- businessMatching
+          ba <- bm.activities
+        } yield { Ok(what_you_need(index, Some(ba), bm.msbServices))
+      })getOrElse(InternalServerError("Unable to retrieve correct data"))
```
Comment (ajchilts) [inline app/controllers/tradingpremises/WhatYouNeedController.scala:42]: Why has the error message changed here? Should this not be "Unable to retrieve business activities" for consistency?

File: app/views/renewal/what_you_need.scala.html (hunk 5 above)
Comment (ajchilts) [reply app/views/renewal/what_you_need.scala.html:59]: I have provided an example above. Only one call to hasBusinessOrAdditionalActivity and simplified `if` statement.

File: conf/messages (hunk 7 above)
Comment (ajchilts) [reply conf/messages:1416]: Should it not be **Agent’s** ?

File: app/controllers/renewal/WhatYouNeedController.scala [hunk 20]
```diff
@@ -15,29 +15,45 @@
...
-        case Section(_,NotStarted | Started,_,_) => Ok(what_you_need())
-        case _ => Redirect(controllers.routes.RegistrationProgressController.get())
-      }
+        (for {
+          bm <- OptionT(dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key))
+          ba <- OptionT.fromOption[Future](bm.activities)
+          section <- OptionT.liftF(getSection(renewalService, request.credId, Some(ba), bm.msbServices))
+        } yield {
+          section
+        }).getOrElse(InternalServerError("Unable to retrieve the correct data"))
```
Comment (ajchilts) [inline app/controllers/renewal/WhatYouNeedController.scala:48]: Why has the error message changed here? Should this not be "Unable to retrieve business activities" for consistency?

File: conf/messages (hunk 7 above)
Comment (ajchilts) [reply conf/messages:1416]: If this is me not being grammatically correct I am sorry. I will check with Lisa. :-)

File: app/controllers/tradingpremises/WhatYouNeedController.scala (hunk 19 above)
Comment (matthewharris1) [reply app/controllers/tradingpremises/WhatYouNeedController.scala:42]: Done

File: app/controllers/renewal/WhatYouNeedController.scala (hunk 20 above)
Comment (matthewharris1) [reply app/controllers/renewal/WhatYouNeedController.scala:48]: Done

**Comment by ajchilts** [comment]
@mostafasina  - merging this PR as the Jira ticket is in done. 

https://jira.tools.tax.service.gov.uk/browse/AMLS-5877

**Pull request merged**